examples/04-shadowing.md

```diff
@@ -0,0 +1,52 @@
+# Shadowing
+
+Module declarations can be shadowed by other modules declarations declared in nested modules:
+
+```js
+module foo { export const x = 1; }
+
+module bar {
+    module foo { export const x = 2; }
+
+    import { x } from foo;
+    console.log(x); // 2
+}
+
+import { x } from foo;
+console.log(x); // 1
+```
+
+Module declarations can also be shadowed by other variables: in that case they cannot be imported anymore.
```

Comment:
how, if you can't reassign module declarations, and module declarations are hoisted?

Reply:
shadowed like this:

```js
module x {
    let x = 1
    import x // error! shadowed by local declaration x
}

{
    let x;
    import(x) // error! undefined is not a valid specifier
}
import x // ok
```

Reply:
Makes sense, but I’m not sure if this example matches fast.

Reply:
The example below is identical to this one, except that:
- I used different variable/module names
- I used two modules (`foo` and `nested`) instead of one (`x`), to avoid re-using the same one for two purposes (declaring the module, and trying to statically import a shadowed module).

Reply:
Oh maybe the confusion comes from different meanings of hoisting.

Not hoisted as in "available in an outer scope":
```js
{
  var x = 1;
}
x === 1;
```
but as in "available before their definition":
```js
f();

function f() {}
```

Reply:
Why would we want to add another construct to the language that’s available before it’s definition? That’s why ES6 has the TDZ for let/const imo - to avoid the mistake of var and function declarations.

Reply:
- `var` hoisting is problematic because the declaration is hoisted, but its value is not and thus you can read the variable but it's not really usable.
- `function` hoisting is not considered problematic in general: some coding styles prefer to keep functions at the bottom, in a "the more you scroll the more internal details you see" pattern. It causes problems when the function captures a variable from the outer scope, and the function is called before that that variable is initialized. Module declarations don't have this problem, because they don't capture variables.
- `class` didn't follow the function pattern because, as noted by @Jack-Works in https://github.com/tc39/proposal-module-fragments/pull/21#discussion_r1008624192, class declarations execute some code when evaluated and thus they cannot be used before evaluating the declaration.

Both TDZ and not TDZ would be fine for me: I'll open an issue to discuss about it as soon as this is merged, since with just two reviews I got two different opinions :stuck_out_tongue:

Reply:
I don't agree that *relying on* function hoisting isn't considered problematic in general - there's coding styles for everything, but that doesn't mean everything's fine.